torchgeo/datasets/ssl4eo.py

```diff
@@ -0,0 +1,231 @@
+        """
+        directory = os.path.join(self.root, self.split, f"{index // self.times:07}")
+        subdirs = sorted(os.listdir(directory))
+        directory = os.path.join(directory, subdirs[index % self.times])
+
+        images = []
+        for band in self.metadata[self.split]["bands"]:
+            filename = os.path.join(directory, f"{band}.tif")
+            with rasterio.open(filename) as f:
+                image = f.read(out_shape=(1, self.size, self.size)).astype(np.float32)
+                images.append(torch.from_numpy(image))
+
+        sample = {"image": torch.cat(images)}
+
+        if self.transforms is not None:
+            sample = self.transforms(sample)
+
+        return sample
+
+    def __len__(self) -> int:
+        """Return the number of data points in the dataset.
+
+        Returns:
+            length of the dataset
+        """
+        return self.locations * self.times
+
+    def _verify(self) -> None:
+        """Verify the integrity of the dataset.
+
+        Raises:
+            RuntimeError: if dataset is missing or checksum fails
+        """
+        # Check if the extracted files already exist
+        directory_path = os.path.join(self.root, self.split)
+        if os.path.exists(directory_path):
+            return
+
+        # Check if the zip files have already been downloaded
+        filename = cast(str, self.metadata[self.split]["filename"])
+        zip_path = os.path.join(self.root, filename)
+        md5 = self.metadata[self.split]["md5"] if self.checksum else None
+        integrity = check_integrity(zip_path, md5)
+        if integrity:
+            self._extract()
+        else:
+            raise RuntimeError(f"Dataset not found in `root={self.root}`")
+
+    def _extract(self) -> None:
+        """Extract the dataset."""
+        filename = cast(str, self.metadata[self.split]["filename"])
+        extract_archive(os.path.join(self.root, filename))
+
+    def plot(
+        self,
+        sample: Dict[str, Tensor],
+        show_titles: bool = True,
+        suptitle: Optional[str] = None,
+    ) -> plt.Figure:
+        """Plot a sample from the dataset.
+
+        Args:
+            sample: a sample returned by :meth:`__getitem__`
+            show_titles: flag indicating whether to show titles above each panel
+            suptitle: optional string to use as a suptitle
+
+        Returns:
+            a matplotlib Figure with the rendered sample
+        """
+        if self.split == "s1":
+            # See Sentinel1.plot
+            co_polarization = sample["image"][0]
+            cross_polarization = sample["image"][1]
+            ratio = co_polarization / cross_polarization
+
+            co_polarization = torch.clamp(co_polarization / 0.3, min=0, max=1)
+            cross_polarization = torch.clamp(cross_polarization / 0.05, min=0, max=1)
+            ratio = torch.clamp(ratio / 25, min=0, max=1)
```

Comment:
@RitwikGupta this plotting trick isn't working, all I see is a black image with red specs. I know you're not the biggest fan of RGB plotting, but this dataset presumably only has one type of image processing so it should be possible to reliably plot. Not sure how to tell if the image is power/decibel/amplitude or gamma/sigma and what that actually means for plotting. Any advice? You can download the [100 patch subset](https://github.com/zhu-xlab/SSL4EO-S12) if you want to test the plotting method.

Reply:
@adamjstewart The Sentinel-1 imagery they pulled from GEE is in the decibel scale. Normally, Sentinel-1 data from Copernicus is in the power scale. The ratio you want to calculate should be in the power scale. This code will work to do visualize this specific data:

```
cross_pol = np.exp(cross_pol / 10)
co_pol = np.exp(co_pol / 10)
ratio = cross_pol / co_pol

fig, ax = plt.subplots(1, 4, figsize=(20, 10))
ax[0].imshow(cross_pol, cmap='bone')
ax[1].imshow(co_pol, cmap='bone')
ax[2].imshow(ratio, cmap='bone')
ax[3].imshow(np.concatenate([cross_pol, co_pol, ratio], axis=2))
```
<img width="1437" alt="image" src="https://user-images.githubusercontent.com/4959141/222085186-dc37e61b-87f5-4815-b494-952062eb205c.png">

Reply:
Here's the relevant GEE documentation that says they convert to decibel when pulling S1 data in GEE: https://developers.google.com/earth-engine/datasets/catalog/COPERNICUS_S1_GRD
<img width="729" alt="image" src="https://user-images.githubusercontent.com/4959141/222085596-093374a6-381f-4d7a-80f6-f23df88acbc8.png">

Reply:
1. I thought it was supposed to be [co_pol, cross_pol, ratio], not [cross_pol, co_pol, ratio]?
2. Am I not supposed to do [normalization](https://gis.stackexchange.com/a/400780/123758) (co_pol / 0.3, cross_pol / 0.05, ratio / 25)?
3. Even with divide by 10, no normalization, swap co_pol/cross_pol, the images don't look great:
![Figure_1](https://user-images.githubusercontent.com/12021217/222225013-b311eb00-b79e-4f9e-a5cb-a307742b4f8f.png)

Reply:
Which image are you using? I can test with the same one.
* I think you can keep [co_pol, cross_pol, ratio].
* Are you doing `np.exp(x / 10)` or just `x /10`?
* You should use the log scale values in that post for viz purposes only. It's not required, but it may help. Do min-max norm to those ranges, not just division.

Reply:
why do you want to use power values to plot? In my paper, I simply plot the dB values (99% percentile and scale each one to [0,1])

Reply:
then can also plot dB for each channel? e.g. [VV,VH,VV/VH]?

Reply:
I think that's what I was doing originally and it didn't work either. I just copied the plotting code from our `Sentinel1` dataset, which worked fine for data downloaded from ASF DAAC, but apparently not for data downloaded from GEE.